### PR TITLE
Extract out `coin-selection` and `primitive` libraries.

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue number

ADP-2386

## Summary

This PR extracts out the following libraries:

- `cardano-wallet-coin-selection`
- `cardano-wallet-primitive`

## Details

- The `cardano-wallet-coin-selection` library:
    - contains modules that relate to coin selection algorithms.
    - can eventually be imported by our `balance-transaction` library.
- The `cardano-wallet-primitive` library:
    - contains just the **_subset_** of `Primitive` modules on which both `cardano-wallet` and `cardano-wallet-coin-selection` currently depend.
    - does **_not_** contain **_all_** modules from the `Primitive` hierarchy, as this hierarchy has a very large transitive dependency footprint.

## Testing

- `cardano-wallet-coin-selection`
    - tested by hydra [here](https://hydra.iohk.io/build/21719506) ([log](https://hydra.iohk.io/build/21719506/log))
- `cardano-wallet-primitive`
    - tested by hydra [here](https://hydra.iohk.io/build/21721055) ([log](https://hydra.iohk.io/build/21721055/log))

## Direction

We intend to steadily **_reduce_** the size of the `cardano-wallet-primitive` library and eventually **_eliminate_** it, by revising `cardano-wallet-coin-selection` (and `balance-transaction`) so that it does not depend on these types.

### CODE REVIEW

**Comment by jonathanknowles** [comment]
bors try

File: lib/coin-selection/cardano-wallet-coin-selection.cabal [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    ghc-options: -O2 -Werror
+
+flag release
+  description: Enable optimization and `-Werror`
+  default:     False
+  manual:      True
+
+flag scrypt
+  description: Enable compatibility support for legacy wallet passwords.
+  default:     True
```
Comment (Anviking) [inline lib/coin-selection/cardano-wallet-coin-selection.cabal:39]: It seems this shouldn't be needed in a coin-selection library.

File: lib/primitive/cardano-wallet-primitive.cabal [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    ghc-options: -O2 -Werror
+
+flag release
+  description: Enable optimization and `-Werror`
+  default:     False
+  manual:      True
+
+flag scrypt
+  description: Enable compatibility support for legacy wallet passwords.
+  default:     True
```
Comment (Anviking) [inline lib/primitive/cardano-wallet-primitive.cabal:39]:
```suggestion
```

File: lib/primitive/cardano-wallet-primitive.cabal [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+library
+  import:          language, opts-lib
+  hs-source-dirs:  lib
+
+  if flag(scrypt)
+    cpp-options:   -DHAVE_SCRYPT
+    build-depends: scrypt
```
Comment (Anviking) [inline lib/primitive/cardano-wallet-primitive.cabal:47]:
```suggestion
```

File: lib/coin-selection/cardano-wallet-coin-selection.cabal [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+library
+  import:          language, opts-lib
+  hs-source-dirs:  lib
+
+  if flag(scrypt)
+    cpp-options:   -DHAVE_SCRYPT
+    build-depends: scrypt
```
Comment (Anviking) [inline lib/coin-selection/cardano-wallet-coin-selection.cabal:47]:
```suggestion
```

**Review by Anviking — APPROVED** [review]
Awesome, lgtm!

File: lib/coin-selection/cardano-wallet-coin-selection.cabal (hunk 1 above)
Comment (jonathanknowles) [reply lib/coin-selection/cardano-wallet-coin-selection.cabal:39]: > It seems this shouldn't be needed in a coin-selection library.

I **_think_** you're probably right, that it's not needed.

If it's okay with you, I'll make another PR immediately after this one that fixes this (and also the other suggestions you've made).

Thank-you for spotting this!

File: lib/primitive/cardano-wallet-primitive.cabal (hunk 2 above)
Comment (jonathanknowles) [reply lib/primitive/cardano-wallet-primitive.cabal:39]: https://github.com/input-output-hk/cardano-wallet/pull/3591#discussion_r1026327817

File: lib/primitive/cardano-wallet-primitive.cabal (hunk 3 above)
Comment (jonathanknowles) [reply lib/primitive/cardano-wallet-primitive.cabal:47]: https://github.com/input-output-hk/cardano-wallet/pull/3591#discussion_r1026327817

File: lib/coin-selection/cardano-wallet-coin-selection.cabal (hunk 4 above)
Comment (jonathanknowles) [reply lib/coin-selection/cardano-wallet-coin-selection.cabal:47]: https://github.com/input-output-hk/cardano-wallet/pull/3591#discussion_r1026327817

**Comment by jonathanknowles** [comment]
It looks like the failure [above](https://github.com/input-output-hk/cardano-wallet/pull/3591#issuecomment-1319932086) is indeed reproducible on `master` (a5fc0eb4e3357739f6de9488169832f33812a34f) (i.e., not caused by this PR):
Screenshot from 2022-11-18 20-41-01

**Comment by jonathanknowles** [comment]
I'm going to merge this PR manually, since:
- We already have one successful `bors try` (https://github.com/input-output-hk/cardano-wallet/pull/3591#issuecomment-1319509062)
- The last run of `bors r+` (https://github.com/input-output-hk/cardano-wallet/pull/3591#issuecomment-1319932086) failed with one error, but that error is reproducible on `master`. (https://github.com/input-output-hk/cardano-wallet/pull/3591#issuecomment-1319947127)
- I have confirmed that all tests pass on my local machine.

File: lib/coin-selection/cardano-wallet-coin-selection.cabal (hunk 1 above)
Comment (jonathanknowles) [reply lib/coin-selection/cardano-wallet-coin-selection.cabal:39]: @Anviking https://github.com/input-output-hk/cardano-wallet/pull/3596

**Pull request merged**